zilliqa/src/block_store.rs

```diff
@@ -18,18 +19,109 @@ use crate::{
 pub struct BlockStore {
     db: Arc<Db>,
     block_cache: RefCell<LruCache<Hash, Block>>,
+    /// The maximum view of blocks we've seen.
+    highest_known_view: u64,
+    /// The maximum view of blocks we've sent a request for.
+    requested_view: u64,
+    /// The maximum number of blocks to send requests for at a time.
+    max_blocks_in_flight: u64,
+    /// The maximum number of blocks to request at a time.
+    batch_size: u64,
+    /// Buffered block proposals, indexed by their parent hash.
+    //buffered: HashMap<Hash, Proposal>,
+    buffered: LruCache<Hash, Proposal>,
     message_sender: MessageSender,
 }
 
 impl BlockStore {
-    pub fn new(db: Arc<Db>, message_sender: MessageSender) -> Result<Self> {
+    pub fn new(config: &NodeConfig, db: Arc<Db>, message_sender: MessageSender) -> Result<Self> {
         Ok(BlockStore {
             db,
             block_cache: RefCell::new(LruCache::new(NonZeroUsize::new(5).unwrap())),
+            highest_known_view: 0,
+            requested_view: 0,
+            max_blocks_in_flight: config.max_blocks_in_flight,
+            batch_size: config.block_request_batch_size,
+            buffered: LruCache::new(
+                NonZeroUsize::new(config.max_blocks_in_flight as usize + 100).unwrap(),
+            ),
             message_sender,
         })
     }
 
+    /// Buffer a block proposal whose parent we don't yet know about.
+    pub fn buffer_proposal(
+        &mut self,
+        proposal: Proposal,
+        random_peer: Option<PeerId>,
+    ) -> Result<()> {
+        let view = proposal.view();
+
+        // We want to be careful about buffering too many proposals. There is no guarantee or proof that any of them
+        // will eventually form our canonical chain. Therefore we limit the number of buffered proposals to
+        // `max_blocks_in_flight + 100`. This number is chosen because it guarantees we will always be able to handle
+        // pending requests made by ourselves that arrive out-of-order, while also giving some extra space for newly
+        // created blocks that arrive while we are syncing.
```

Comment:
Given that we can have an arbitrary number of proposals from other nodes, why does '+100' guarantee that we will always be able to handle pending requests made by us?

Reply:
It doesn't and "guarantee" is a silly word, thanks for spotting :)